### PR TITLE
add packer optimized by `numba`

### PR DESCRIPTION
**Pull request by yiliu30**

### CODE REVIEW

File: auto_round/utils.py [hunk 1]
```diff
@@ -664,3 +664,161 @@ def dynamic_import_inference_linear(backend, bits, group_size, sym):
...
+        new_in_features = (in_features + n_pack - 1) // n_pack
+        packed_tensor = np.zeros((out_features, new_in_features), dtype=compression_dtype)
+        raw_tensor = raw_tensor.astype(compression_dtype)
+
+        if bits == 4:
+            for i in range(new_in_features):
+                packed_tensor[:, i] = (
+                    (raw_tensor[:, i * n_pack + 7] << 28)
+                    | (raw_tensor[:, i * n_pack + 6] << 24)
+                    | (raw_tensor[:, i * n_pack + 5] << 20)
```
Comment (wenhuach21) [inline auto_round/utils.py:721]: is the packing logic same with autogptq exllama? better integrate it into one format and have a accuracy check

**Comment by Qubitium** [comment]
@yiliu30 How much faster is this packer vs the current packer? I assume exllama and triton with triton being the only one that supports 2bits.

**Comment by yiliu30** [comment]
> @yiliu30 How much faster is this packer vs the current packer? I assume exllama and triton with triton being the only one that supports 2bits.

Hi @Qubitium, Since this PR is still a work in progress, I have not compared it with [`qlinear_triton`'s pack](https://github.com/intel/auto-round/blob/0daad3f26b158e1f51dad233722fe760c7069a99/auto_round_extension/cuda/qliner_triton.py#L130). However, there is a reference implementation (`pack_2d_tensor`) that is similar to `qlinear_triton`'s `pack`. I compare it with the optimized pack(`pack_tensor_with_numpy_opt_np_numba_v2`). 

The speedup depends on the tensor shape and hardware. You can test it on your side using the command below if you are interested.

```bash
BENCH_PACKER=1 numactl -l -C 0-11 pytest -sv ./test/test_utility.py -k test_pack
```
Here are some test results on my side.
```bash
test/test_utility.py::test_pack[1024-128-2] ref : 0.0874624450225383,  res: 2.2586073159764055, speed up: 0.038724059912436756
PASSED
test/test_utility.py::test_pack[1024-128-4] ref : 0.09346804500091821,  res: 0.0701686849934049, speed up: 1.332047836006948
PASSED
test/test_utility.py::test_pack[1024-1024-2] ref : 0.15939546999288723,  res: 0.04957061700406484, speed up: 3.2155232197294747
PASSED
test/test_utility.py::test_pack[1024-1024-4] ref : 0.1800197789852973,  res: 0.086655695020454, speed up: 2.077414288152739
PASSED
test/test_utility.py::test_pack[1024-5120-2] ref : 0.4963021900039166,  res: 0.07299031602451578, speed up: 6.799562147904937
PASSED
test/test_utility.py::test_pack[1024-5120-4] ref : 0.5322486590011977,  res: 0.12510409200331196, speed up: 4.25444644118521
PASSED
test/test_utility.py::test_pack[1024-13824-2] ref : 1.4805711419903673,  res: 0.11439396301284432, speed up: 12.942738436504088
PASSED
test/test_utility.py::test_pack[1024-13824-4] ref : 1.6657754300103988,  res: 0.19896381799480878, speed up: 8.372253039765557
PASSED
test/test_utility.py::test_pack[13824-128-2] ref : 1.1718379610101692,  res: 0.501374389976263, speed up: 2.3372513324137847
PASSED
test/test_utility.py::test_pack[13824-128-4] ref : 1.2479936860036105,  res: 0.9018774160067551, speed up: 1.3837730758680655
PASSED
test/test_utility.py::test_pack[13824-1024-2] ref : 2.2658500450197607,  res: 0.6737105899956077, speed up: 3.3632394661252585
PASSED
test/test_utility.py::test_pack[13824-1024-4] ref : 2.530184699979145,  res: 1.1530198710097466, speed up: 2.1943981743899688
PASSED
test/test_utility.py::test_pack[13824-5120-2] ref : 10.542667002009694,  res: 1.4736216420133132, speed up: 7.154256358237203
PASSED
test/test_utility.py::test_pack[13824-5120-4] ref : 12.083586978988023,  res: 2.273310685006436, speed up: 5.315413796576517
PASSED
test/test_utility.py::test_pack[13824-13824-2] ref : 27.6478629320045,  res: 2.663787511002738, speed up: 10.379154800375547
PASSED
test/test_utility.py::test_pack[13824-13824-4] ref : 32.4161597469938,  res: 3.7381430180103052, speed up: 8.671728072150618
PASSED
```

**Comment by wenhuach21** [comment]
no update for a long time

**Pull request closed without merging**